### PR TITLE
🐛 Fix the deletion step order and OwnerReference matching, add matchpolicy to webhooks

### LINKED ISSUE #22: capm3-controller-manager invalid memory address error
**What steps did you take and what happened:**
Provision and de-provision a control plane machine multiple times.
It goes to CrashLoopBackOff with invalid memory address error

full log is added as a comment
**What did you expect to happen:**
A success or a graceful failure.

**Anything else you would like to add:**
Regardless of the cause, please do focus on fixing the ` invalid memory address error`

This makes the environment not usable. One needs to run `make clean` and `make` again.
**Environment:**

- Cluster-api version: v1alpha3
- CAPBM version: v1alpha3
- Minikube version: v1.8.2
- environment: metal3-dev-env
- Kubernetes version: 1.17.0

/kind bug

### LINKED ISSUE #30: CAPM3 should not set error on transient failures
**What steps did you take and what happened:**
For example when failing to update the BMH due to a conflict, CAPM3 reports an error on the Metal3Machine. CAPI picks it up : 

```
apiVersion: v1
items:
- apiVersion: cluster.x-k8s.io/v1alpha3
  kind: Machine
  metadata:
    creationTimestamp: "2020-03-22T08:26:40Z"
    finalizers:
    - machine.cluster.x-k8s.io
    generation: 3
    labels:
      cluster.x-k8s.io/cluster-name: test1
      cluster.x-k8s.io/control-plane: ""
      kubeadm.controlplane.cluster.x-k8s.io/hash: "3288331389"
    name: test1-controlplane-lpnxq
    namespace: metal3
    ownerReferences:
    - apiVersion: controlplane.cluster.x-k8s.io/v1alpha3
      blockOwnerDeletion: true
      controller: true
      kind: KubeadmControlPlane
      name: test1-controlplane
      uid: 834ec5bd-3cbf-4b87-9ccc-f8b5b07140d4
    resourceVersion: "6836"
    selfLink: /apis/cluster.x-k8s.io/v1alpha3/namespaces/metal3/machines/test1-controlplane-lpnxq
    uid: 6be1adb7-c185-4779-be4c-d04b0e396f2e
  spec:
    bootstrap:
      configRef:
        apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
        kind: KubeadmConfig
        name: test1-controlplane-bzpmb
        namespace: metal3
        uid: 5c3bbc07-b8f3-4ccb-98f5-be299edf226e
      dataSecretName: test1-controlplane-bzpmb
    clusterName: test1
    infrastructureRef:
      apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
      kind: Metal3Machine
      name: test1-controlplane-jfnpf
      namespace: metal3
      uid: 1cc69570-8c3d-4fbd-a8d6-c7d25118cb31
    providerID: metal3://d99b9c2b-5aee-45e6-9249-e7b42f7ccc76
    version: v1.17.0
  status:
    addresses:
    - address: 192.168.111.21
      type: InternalIP
    - address: 172.22.0.91
      type: InternalIP
    - address: node-1
      type: Hostname
    - address: node-1
      type: InternalDNS
    bootstrapReady: true
    failureMessage: 'Failure detected from referenced resource infrastructure.cluster.x-k8s.io/v1alpha3,
      Kind=Metal3Machine with name "test1-controlplane-jfnpf": Failed to associate
      the BaremetalHost to the Metal3Machine'
    failureReason: CreateError
    infrastructureReady: true
    lastUpdated: "2020-03-22T08:34:41Z"
    nodeRef:
      name: node-1
      uid: 49aa2fb7-b645-4da3-9600-64e31074bf81
    phase: Failed
- apiVersion: cluster.x-k8s.io/v1alpha3
  kind: Machine
  metadata:
    creationTimestamp: "2020-03-22T08:26:21Z"
    finalizers:
    - machine.cluster.x-k8s.io
    generateName: test1-md-0-78d55dc456-
    generation: 3
    labels:
      cluster.x-k8s.io/cluster-name: test1
      machine-template-hash: "3481187012"
      nodepool: nodepool-0
    name: test1-md-0-78d55dc456-5b86v
    namespace: metal3
    ownerReferences:
    - apiVersion: cluster.x-k8s.io/v1alpha3
      blockOwnerDeletion: true
      controller: true
      kind: MachineSet
      name: test1-md-0-78d55dc456
      uid: 7b2e3dbc-8af2-4f43-8945-6968bc12bd76
    resourceVersion: "9982"
    selfLink: /apis/cluster.x-k8s.io/v1alpha3/namespaces/metal3/machines/test1-md-0-78d55dc456-5b86v
    uid: 3826ad1f-b1ed-4504-b265-2dde934055e9
  spec:
    bootstrap:
      configRef:
        apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
        kind: KubeadmConfig
        name: test1-md-0-5hhsj
        namespace: metal3
        uid: 2d396de3-c3f4-45cc-baff-d915ebcfb20d
      dataSecretName: test1-md-0-5hhsj
    clusterName: test1
    infrastructureRef:
      apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
      kind: Metal3Machine
      name: test1-md-0-hr5jw
      namespace: metal3
      uid: 3477a996-d531-4901-a20e-bd9c93b1a8e6
    providerID: metal3://6fc41e7b-cc2b-4a06-a9d4-883c563d0c47
    version: v1.17.0
  status:
    addresses:
    - address: 172.22.0.87
      type: InternalIP
    - address: 192.168.111.20
      type: InternalIP
    - address: node-0
      type: Hostname
    - address: node-0
      type: InternalDNS
    bootstrapReady: true
    failureMessage: 'Failure detected from referenced resource infrastructure.cluster.x-k8s.io/v1alpha3,
      Kind=Metal3Machine with name "test1-md-0-hr5jw": Failed to associate the BaremetalHost
      to the Metal3Machine'
    failureReason: CreateError
    infrastructureReady: true
    lastUpdated: "2020-03-22T08:45:24Z"
    nodeRef:
      name: node-0
      uid: 8d146172-c593-455e-bf3c-c5c1dcb1bb73
    phase: Failed
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**What did you expect to happen:**
Since it is a temporary failure, waiting for a requeue, it should not be set as error.

**Environment:**

- Cluster-api version: v1alpha3
- CAPBM version: v1alpha3
- Minikube version: NA
- environment (metal3-dev-env or other): NA
- Kubernetes version: (use `kubectl version`): NA

/kind bug

### PR DESCRIPTION
**Pull request by maelk**
**What this PR does / why we need it**:
This PR fixes the order of the deletion steps for Metal3Machine to prevent a NPE (#22)
It also changes the way the OwnerReferences are matched to not take into account the version, that
might change after an upgrade.
It also upgrades controller-gen for CAPI v0.3.2
It adds matchPolicy=Equivalent to the webhook to match across versions

**Which issue(s) this PR fixes**:
Fixes #22 
Fixes #30

### CODE REVIEW

File: baremetal/metal3machine_manager.go [hunk 1]
```diff
@@ -105,7 +106,7 @@ func NewMachineManager(client client.Client,
 // SetFinalizer sets finalizer
 func (m *MachineManager) SetFinalizer() {
 	// If the Metal3Machine doesn't have finalizer, add it.
-	if !util.Contains(m.Metal3Machine.Finalizers, capm3.MachineFinalizer) {
+	if !Contains(m.Metal3Machine.Finalizers, capm3.MachineFinalizer) {
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go]: What is the motivation behind not using the util functions and writing own?

Comment (maelk) [reply baremetal/metal3machine_manager.go]: Cluster API removed the Contains and Filter functions from their util package in the latest version

File: baremetal/metal3machine_manager.go [hunk 2]
```diff
@@ -926,9 +940,12 @@ func (m *MachineManager) SetProviderID(providerID string) {
 }
 
 // SetOwnerRef adds an ownerreference to this Metal3 machine
-func (m *MachineManager) SetOwnerRef(refList []metav1.OwnerReference, controller bool) []metav1.OwnerReference {
+func (m *MachineManager) SetOwnerRef(refList []metav1.OwnerReference, controller bool) ([]metav1.OwnerReference, error) {
 	index, err := m.FindOwnerRef(refList)
 	if err != nil {
+		if _, ok := err.(*NotFoundError); !ok {
```
[the remark this answers is not in the document]
Comment (maelk) [reply baremetal/metal3machine_manager.go]: I am not sure to get what you mean

[the remark this answers is not in the document]
Comment (kashifest) [reply baremetal/metal3machine_manager.go]: I didn't realize it was a custom error type. Ignore my comment

File: baremetal/metal3machine_manager.go [hunk 3]
```diff
@@ -453,64 +457,68 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
 		if host.Labels != nil && host.Labels[capi.ClusterLabelName] == m.Machine.Spec.ClusterName {
 			delete(host.Labels, capi.ClusterLabelName)
 		}
 
-		err = m.updateObject(ctx, host)
-		if err != nil && !apierrors.IsNotFound(err) {
-			if _, ok := err.(HasRequeueAfterError); !ok {
-				m.setError("Failed to delete Metal3Machine",
+		// Delete created secret, if data was set without DataSecretName or if
+		// BareMetalHost and Machine are in different namespaces.
```
[the remark this answers is not in the document]
Comment (maelk) [reply baremetal/metal3machine_manager.go]: By default they watch all namespaces. And rigth now, the scenario is not possible. I have created an issue yesterday : https://github.com/metal3-io/cluster-api-provider-metal3/issues/29

**Comment by maelk** [comment]
> I find it confusing that 'baremetal' is still used in parameter names and directory or file names.

We will anyways not solve this on this PR. Maybe a next one

**Comment by maelk** [comment]
> looks good to me, but the tiny nit, why in some places we still have BM* instead of M3*?

We could fix this. Maybe in another PR ?

**Pull request merged**